core/notification/src/main/kotlin/edu/stanford/spezi/core/notification/fcm/DeviceRegistrationService.kt

```diff
@@ -3,96 +3,120 @@ package edu.stanford.spezi.core.notification.fcm
+        logger.i { "Received register device request: $token" }
+        if (token == getStorageToken()) {
+            logger.i { "Ignoring registerDevice as the same token has been already submitted" }
+            return
+        }
         val platform = PLATFORM_ANDROID
         val osVersion = Build.VERSION.RELEASE
-        val packageInfo = getPackageInfo()
+        val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
         val language = Locale.getDefault().toLanguageTag()
         val timeZone = TimeZone.getDefault().id
-
-        val deviceInfo = DeviceInfo(
-            notificationToken = token,
-            platform = platform,
-            osVersion = osVersion,
-            appVersion = packageInfo.versionName,
-            appBuild = packageInfo.versionCode.toString(),
-            language = language,
-            timeZone = timeZone
+        val body = mapOf(
+            KEY_NOTIFICATION_TOKEN to token,
+            KEY_PLATFORM to platform,
+            KEY_OS_VERSION to osVersion,
+            KEY_APP_VERSION to packageInfo.versionName,
+            KEY_APP_BUILD to packageInfo.versionCode.toString(),
+            KEY_LANGUAGE to language,
+            KEY_TIME_ZONE to timeZone
         )
-
-        sendDeviceInfoToServer(deviceInfo)
+        coroutineScope.launch {
+            runCatching {
+                functions.getHttpsCallable(REGISTER_DEVICE_FUNCTION).call(body).await()
+                storage.putString(key = STORAGE_KEY_NOTIFICATION_TOKEN, value = token)
+                logger.i { "Successfully registered device: $body" }
+            }.onFailure { e ->
+                logger.e(e) { "Exception occurred while registering device" }
+            }
+        }
     }
 
-    private fun getPackageInfo() = context.packageManager.getPackageInfo(context.packageName, 0)
-
-    private fun sendDeviceInfoToServer(deviceInfo: DeviceInfo) {
-        val deviceInfoMap = with(deviceInfo) {
-            mapOf(
-                KEY_NOTIFICATION_TOKEN to notificationToken,
-                KEY_PLATFORM to platform,
-                KEY_OS_VERSION to osVersion,
-                KEY_APP_VERSION to appVersion,
-                KEY_APP_BUILD to appBuild,
-                KEY_LANGUAGE to language,
-                KEY_TIME_ZONE to timeZone
-            )
+    override fun refreshDeviceToken() {
+        logger.i { "received refreshDeviceToken request" }
+        coroutineScope.launch {
+            runCatching {
+                val token = firebaseMessaging.token.await()
+                logger.i { "registering device with fm token: $token" }
+                registerDevice(token = token)
+            }.onFailure { error ->
+                logger.e(error) { "Error while accessing firebase messaging token" }
+            }
         }
+    }
 
+    override suspend fun unregisterDevice() {
+        logger.i { "Received unregister device request" }
+        val storageToken = getStorageToken() ?: run {
```

Comment:
If I see this correctly, then we wouldn't actually need to store the notification token, if there is a call (`firebaseMessaging.token`) to get the current token, right? Or is there harm in calling that function in some cases (e.g. triggering permission alerts or sth like that)?

The firebase function simply ignores, if a token is not registered already, so it wouldn't actually make a big difference, if we call it with an unregistered token (except environmentally, maybe...)

Reply:
Nope, there is no harm to call that token method. it is just a blocking call and persisting locally is more lightweight compared to that call. furthermore, local storage is proposed mechanism to keep track token management of the device, e.g. to reduce number of unnecessary remote calls. you can read more in the [firebase docs](https://firebase.google.com/docs/cloud-messaging/manage-tokens) and in this firebase sample [here](https://github.com/firebase/quickstart-android/blob/0b77e46545ef38b040d140f360cfe3f75b54e15d/messaging/app/src/main/java/com/google/firebase/quickstart/fcm/kotlin/MainActivity.kt#L131-L153)